### PR TITLE
Save filters & enhancement for 3D LUT files

### PR DESCRIPTION
**Pull request by lhc70000**
**Description:**

- Users can now save an added filter, and assign a shortcut key to it.
- Support dragging a 3D LUT file (3dl, cube, dat or m3d) to main window to apply it directly.

### CODE REVIEW

File: iina/FilterPresets.swift [hunk 1]
```diff
@@ -69,38 +69,47 @@ class FilterPresetInstance {
...
-    let pm = FilterParameter(.float, defaultValue: FilterParamaterValue(float: defaultValue))
+    let pm = FilterParameter(.float, defaultValue: FilterParameterValue(float: defaultValue))
     pm.min = min
     pm.max = max
     return pm
   }
 
-  private init(_ type: ParamType, defaultValue: FilterParamaterValue) {
+  static func choose(from choices: [String], defaultChoiceIndex: Int = 0) -> FilterParameter {
+    guard !choices.isEmpty else { fatalError("FilterParameter: Choices annot be empty") }
```
Comment (saagarjha) [inline iina/FilterPresets.swift]: typo

File: iina/FilterWindowController.swift [hunk 2]
```diff
@@ -62,51 +111,160 @@ class FilterWindowController: NSWindowController {
...
   }
 
   func tableView(_ tableView: NSTableView, objectValueFor tableColumn: NSTableColumn?, row: Int) -> Any? {
-    guard let filter = filters.at(row) else { return nil }
-    if tableColumn?.identifier == .key {
-      return row.toStr()
-    } else if tableColumn?.identifier == .value {
-      return filter.stringFormat
+    if tableView == currentFiltersTableView {
+      guard let filter = filters.at(row) else { return nil }
```
Comment (saagarjha) [inline iina/FilterWindowController.swift]: You're only using this once, so…

File: iina/FilterWindowController.swift [hunk 3]
```diff
@@ -62,51 +111,160 @@ class FilterWindowController: NSWindowController {
...
+      if tableColumn?.identifier == .key {
+        return row.toStr()
+      } else if tableColumn?.identifier == .value {
+        return filter.stringFormat
```
Comment (saagarjha) [inline iina/FilterWindowController.swift]: …`return filters.at(row)?.toStr()`?

File: iina/KeyCodeHelper.swift [hunk 4]
```diff
@@ -7,7 +7,7 @@
 //
 
 import Foundation
-
+import Carbon
```
Comment (saagarjha) [inline iina/KeyCodeHelper.swift]: Ouch…is there no other way to do this? I'd really not like us to use deprecated APIs.

Comment (lhc70000) [reply iina/KeyCodeHelper.swift]: Unfortunately no…… `UCKeyTranslate` itself is not from Carbon, but in order to get the "key layout data", some HIToolbox APIs are needed.

Comment (saagarjha) [reply iina/KeyCodeHelper.swift]: I guess…

**Comment by dsayed** [comment]
It would be good to add a file picker rather than have to type in the path to the file in the filter.

**Comment by lhc70000** [comment]
@dsayed You can drag & drop the file directly into the player window (to apply the LUT immediately) or the input box (to fill in the path).

**Pull request merged**